src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryTable.cs

```diff
@@ -613,71 +613,53 @@ private static FormulaValue SortOptionSet(List<(DValue<RecordValue> row, Formula
         }
 
-        private static async Task<DValue<RecordValue>> LazyFilterRowAsync(
-           EvalVisitor runner,
-           EvalVisitorContext context,
-           DValue<RecordValue> row,
-           LambdaFormulaValue filter)
+        private static async IAsyncEnumerable<DValue<RecordValue>> FilterRowAsync(EvalVisitor runner, EvalVisitorContext context, DValue<RecordValue> row, LambdaFormulaValue filter, [EnumeratorCancellation] CancellationToken cts = default)
         {
-            SymbolContext childContext;
-
-            // Issue #263 Filter should be able to handle empty rows
-            if (row.IsValue)
-            {
-                childContext = context.SymbolContext.WithScopeValues(row.Value);
-            }
-            else if (row.IsBlank)
-            {
-                childContext = context.SymbolContext.WithScopeValues(RecordValue.Empty());
-            }
-            else
-            {
-                return null;
-            }
+            SymbolContext childContext = null;
+
+            // Issue #263 Filter should be able to handle empty rows
+            childContext = row.IsValue ? context.SymbolContext.WithScopeValues(row.Value)
+                         : row.IsBlank ? context.SymbolContext.WithScopeValues(RecordValue.Empty())
+                         : row.IsError ? context.SymbolContext.WithScopeValues(row.Error)
+                         : null;
+
+            cts.ThrowIfCancellationRequested();
+
+            // Filter evals to a boolean 
+            var result = await filter.EvalInRowScopeAsync(context.NewScope(childContext)).ConfigureAwait(false);
 
-            // Filter evals to a boolean 
-            var result = await filter.EvalInRowScopeAsync(context.NewScope(childContext)).ConfigureAwait(false);
             var include = false;
             if (result is BooleanValue booleanValue)
             {
                 include = booleanValue.Value;
             }
             else if (result is ErrorValue errorValue)
             {
-                return DValue<RecordValue>.Of(errorValue);
+                yield return DValue<RecordValue>.Of(errorValue);
             }
 
             if (include)
             {
-                return row;
+                yield return row;
             }
 
-            return null;
+            yield return null;
         }
 
-        private static async Task<DValue<RecordValue>[]> LazyFilterAsync(
-            EvalVisitor runner,
-            EvalVisitorContext context,
-            IEnumerable<DValue<RecordValue>> sources,
-            LambdaFormulaValue filter,
-            int topN = int.MaxValue)
-        {
-            var results = new List<DValue<RecordValue>>();
-
-            // Filter needs to allow running in parallel. 
+        private static async IAsyncEnumerable<DValue<RecordValue>> FilterAsync(EvalVisitor runner, EvalVisitorContext context, IEnumerable<DValue<RecordValue>> sources, LambdaFormulaValue filter, int topN = int.MaxValue)
```

Comment:
> IAsyncEnumerable<DValue<RecordValue>>  [](http://example.com/codeflow?start=28&length=39)

This is significant:
before, we return T[], which is mutable, (but not Lazy!)
Now we return IEnum which is lazy, but not mutable. 

Mutable determines Patch, Collect, etc.

Reply:
This is the main reason for this change
Capabilities (or interfaces) will have to be managed by the caller